pkg/plugins/gateway/algorithms/prefix_cache.go

```diff
@@ -18,113 +18,205 @@ package routingalgorithms
+		}
 	}
 
-	var matchedPodNames, readyPodNames []string
-	for _, p := range matchedPods {
-		matchedPodNames = append(matchedPodNames, p.Status.PodIP)
+	// no pod with prefix match, as a fallback select pod with least request count
+	if len(matchedPods) == 0 || targetPod == nil {
+		targetPod = selectTargetPodWithLeastRequestCount(p.cache, readyPods)
+		klog.InfoS("prefix_cache_fallback_least_request_count",
+			"request_id", ctx.RequestID,
+			"target_pod", targetPod.Name,
+			"target_pod_ip", targetPod.Status.PodIP,
+			"matched_pods", matchedPods,
+			"pod_request_count", getRequestCounts(p.cache, readyPods))
 	}
-	for _, p := range readyPods {
-		readyPodNames = append(readyPodNames, p.Status.PodIP)
+
+	if len(prefixHashes) > 0 {
+		p.prefixCacheIndexer.AddPrefix(prefixHashes, ctx.Model, targetPod.Name)
 	}
-	klog.InfoS("prefix cache route",
-		"matched_tokens", matchedTokens,
-		"unmatched_tokens", unMatchedTokens,
-		"matched_pods", matchedPodNames,
-		"ready_pods", readyPodNames,
-		"target_pod", targetPod.Status.PodIP)
 
 	ctx.SetTargetPod(targetPod)
 	return ctx.TargetAddress(), nil
 }
+
+func getTargetPodFromMatchedPods(cache cache.Cache, readyPods []*v1.Pod, matchedPods map[string]int) *v1.Pod {
+	var targetPodName string
+	requestCount := []float64{}
+
+	podRequestCount := getRequestCounts(cache, readyPods)
+	for _, cnt := range podRequestCount {
+		requestCount = append(requestCount, float64(cnt))
+	}
+	meanRequestCount := mean(requestCount)
+	stdDevRequestCount := standardDeviation(requestCount)
+
+	podnames := []string{}
+	for podname := range matchedPods {
+		podnames = append(podnames, podname)
+	}
+	rand.Shuffle(len(podnames), func(i, j int) {
+		podnames[i], podnames[j] = podnames[j], podnames[i]
+	})
+
+	// sort pods with decreasing %perfixmatch AND for same %prefixmatch sort by increasing request count
+	sort.SliceStable(podnames, func(i, j int) bool {
+		if matchedPods[podnames[i]] == matchedPods[podnames[j]] {
+			return podRequestCount[podnames[i]] < podRequestCount[podnames[j]]
+		}
+		return matchedPods[podnames[i]] > matchedPods[podnames[j]]
+	})
+
+	// select targetpod with highest %prefixmatch and request_count within stddev
+	for _, podname := range podnames {
+		reqCnt := float64(podRequestCount[podname])
+		if reqCnt <= meanRequestCount+stdDevRequestCount {
+			targetPodName = podname
+			break
+		}
+	}
+	targetPod, _ := utils.FilterPodByName(targetPodName, readyPods)
+	return targetPod
+}
+
+// getTargetPodOnLoadImbalance evaluates if the load is imbalanced based on the abs difference between
+// pods with min and max outstanding request counts
+func getTargetPodOnLoadImbalance(cache cache.Cache, readyPods []*v1.Pod) (*v1.Pod, bool) {
+	targetPods := []string{}
+	minValue := math.MaxInt32
+	maxValue := math.MinInt32
+
+	podRequestCount := getRequestCounts(cache, readyPods)
+	for podname, value := range podRequestCount {
+		if value <= minValue {
+			minValue = value
+			targetPods = append(targetPods, podname)
```

Comment:
in this case, If the podrequest count gives 5,4,3,2,1, your targetPods may get all of them. 

It's better to use a fixed value, and make large request count pod name can be evict. 2 or 3 makes sense (from Power of Two Choices https://smallrye.io/smallrye-stork/1.1.1/load-balancer/power-of-two-choices/)

Reply:
Good catch, I fixed this bug in the least request. I will fix it here as well.